src/shared/metrics.c

```diff
@@ -0,0 +1,268 @@
+
+static int metric_family_build_json_one(sd_varlink *link, const MetricFamily* mf, bool more) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        int r;
+
+        assert(link);
+        assert(mf);
+
+        r = sd_json_buildo(
+                        &v,
+                        SD_JSON_BUILD_PAIR_STRING("name", mf->name),
+                        SD_JSON_BUILD_PAIR_STRING("description", mf->description),
+                        SD_JSON_BUILD_PAIR_STRING("type", metric_family_type_to_string(mf->type)));
+        if (r < 0)
+                return r;
+
+        if (more)
+                return sd_varlink_notify(link, v);
+
+        return sd_varlink_reply(link, v);
+}
+
+int metrics_method_describe(const MetricFamily metric_family_table[], sd_varlink *link, sd_json_variant *parameters, sd_varlink_method_flags_t flags, void *userdata) {
+        int r;
+
+        assert(metric_family_table);
+        assert(link);
+        assert(parameters);
+
+        r = sd_varlink_dispatch(link, parameters, /* dispatch_table= */ NULL, /* userdata= */ NULL);
+        if (r != 0)
+                return r;
+
+        if (!FLAGS_SET(flags, SD_VARLINK_METHOD_MORE))
+                return sd_varlink_error(link, SD_VARLINK_ERROR_EXPECTED_MORE, NULL);
+
+        const MetricFamily *previous = NULL;
+        for (const MetricFamily *mf = metric_family_table; mf && mf->name; mf++) {
+                if (previous) {
+                        r = metric_family_build_json_one(link, previous, /* more= */ true);
+                        if (r < 0)
+                                return log_debug_errno(r, "Failed to describe metric family '%s': %m", previous->name);
+                }
+
+                previous = mf;
+        }
+
+        if (previous) {
+                r = metric_family_build_json_one(link, previous, /* more */ false);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to describe metric family '%s': %m", previous->name);
+        }
+
+        return 0;
+}
+
+int metrics_method_list(const MetricFamily metric_family_table[], sd_varlink *link, sd_json_variant *parameters, sd_varlink_method_flags_t flags, void *userdata) {
+        int r;
+
+        assert(metric_family_table);
+        assert(link);
+        assert(parameters);
+
+        r = sd_varlink_dispatch(link, parameters, /* dispatch_table= */ NULL, /* userdata= */ NULL);
+        if (r != 0)
+                return r;
+
+        if (!FLAGS_SET(flags, SD_VARLINK_METHOD_MORE))
+                return sd_varlink_error(link, SD_VARLINK_ERROR_EXPECTED_MORE, NULL);
+
+        sd_json_variant *previous = NULL;
+        for (const MetricFamily *mf = metric_family_table; mf && mf->name; mf++) {
+                assert(mf->iterate_cb);
+
+                MetricFamilyContext ctx = {
+                        .metric_family = mf,
+                        .link = link,
+                        .previous = previous};
```

Comment:
You can declare this outside of the for loop and only update the `metric_family` field inside the for loop. Then you can get rid of the `previous` variable declaration outside of the for loop.